pkg/backend/backend.go

```diff
@@ -0,0 +1,280 @@
+// Without that issue resolved, the use of generics would produce too much boilerplate and would look much worse than
+// the solution below — I've tried.
+
+func (backend Backend) SendRender(ctx context.Context, request types.RenderRequest, msgCh chan []types.Metric, errCh chan error) {
+	if cap(backend.renderQ) > 0 {
+		backend.renderQ <- &renderReq{
+			RenderRequest: request,
+			Ctx:           ctx,
+			StartTime:     time.Now(),
+			Results:       msgCh,
+			Errors:        errCh,
+		}
+		backend.requestsInQueue.WithLabelValues("render").Inc()
+		backend.enqueuedRequests.WithLabelValues("render").Inc()
+	} else {
+		// This branch is only necessary to satisfy the tests.
+		// After we clean-up and rework the tests to have mock metrics everywhere,
+		// this should be removed.
+		// TODO: Remove after tests cleanup.
+		go func(b Backend) {
+			msg, err := b.Render(ctx, request)
+			if err != nil {
+				errCh <- err
+			} else {
+				msgCh <- msg
+			}
+		}(backend)
+	}
+}
+
+func (backend Backend) SendFind(ctx context.Context, request types.FindRequest, msgCh chan types.Matches, errCh chan error, durationHist *prometheus.HistogramVec) {
+	if cap(backend.findQ) > 0 {
+		backend.findQ <- &findReq{
+			FindRequest: request,
+			Ctx:         ctx,
+			StartTime:   time.Now(),
+			Results:     msgCh,
+			Errors:      errCh,
+		}
+		backend.requestsInQueue.WithLabelValues("find").Inc()
+		backend.enqueuedRequests.WithLabelValues("find").Inc()
+	} else {
+		// This branch is only necessary to satisfy the tests.
+		// After we clean-up and rework the tests to have mock metrics everywhere,
+		// this should be removed.
+		// TODO: Remove after tests cleanup.
+		go func(b Backend) {
+			var t *prometheus.Timer
+			if durationHist != nil {
+				t = prometheus.NewTimer(durationHist.WithLabelValues(b.GetCluster()))
+			}
+			defer func() {
+				if t != nil {
+					t.ObserveDuration()
+				}
+			}()
+
+			msg, err := b.Find(ctx, request)
+			if err != nil {
+				errCh <- err
+			} else {
+				msgCh <- msg
+			}
+		}(backend)
+	}
+}
+
+func (backend Backend) SendInfo(ctx context.Context, request types.InfoRequest, msgCh chan []types.Info, errCh chan error) {
+	if cap(backend.infoQ) > 0 {
+		backend.infoQ <- &infoReq{
+			InfoRequest: request,
+			Ctx:         ctx,
+			StartTime:   time.Now(),
+			Results:     msgCh,
+			Errors:      errCh,
+		}
+		backend.requestsInQueue.WithLabelValues("info").Inc()
+		backend.enqueuedRequests.WithLabelValues("info").Inc()
+	} else {
+		// This branch is only necessary to satisfy the tests.
+		// After we clean-up and rework the tests to have mock metrics everywhere,
+		// this should be removed.
+		// TODO: Remove after tests cleanup.
+		go func(b Backend) {
+			msg, err := b.Info(ctx, request)
+			if err != nil {
+				errCh <- err
+			} else {
+				msgCh <- msg
+			}
+		}(backend)
+	}
+}
+
+func NewMock(cfg mock.Config) Backend {
+	return NewBackend(mock.New(cfg), 0, 0, nil, nil, nil, nil)
```

Comment:
Maybe it's worth adding `NewMockWithNewBackend` (or with a different name) initializing new backend using the new queues, and run the tests for this backend as well so that tests are also run against the new logic. It shouldn't be any different in the test logic, I guess.

Reply:
It is a typical anti-pattern in Go to repeat the package name in the symbols contained inside the package. If I understand correctly, this suggestion would do precisely that, duplicating the indication of a "backend".
If you mean something different, the old name still makes more sense to me, because it has no additional mocking besides the creation of the backend. 

Also, I hope to clean this up after the redesign is done. This is a necessary thing to keep the tests going.

Reply:
Also, let me kindly remind you of the disclaimer in the PR description) Such small things are not worth discussing on the current stage. A lot of the will change or go away.